File: code/nim/rest/server/src/task_server.nim
import jester
import json
import strutils
import times
import tables
import locks
import uuid4
import options

type
  TaskStatus* = enum
    tsPending = "pending"
    tsInProgress = "in-progress"
    tsCompleted = "completed"
    tsCancelled = "cancelled"

  TaskPriority* = enum
    tpLow = "low"
    tpMedium = "medium"
    tpHigh = "high"
    tpUrgent = "urgent"

  Task* = object
    id*: string
    title*: string
    description*: Option[string]
    status*: TaskStatus
    priority*: TaskPriority
    tags*: seq[string]
    assignedTo*: Option[string]
    createdAt*: float
    updatedAt*: float

  CreateTaskRequest* = object
    title*: string
    description*: Option[string]
    priority*: Option[string]
    tags*: Option[seq[string]]
    assignedTo*: Option[string]

  UpdateTaskRequest* = object
    title*: Option[string]
    description*: Option[string]
    status*: Option[string]
    priority*: Option[string]
    tags*: Option[seq[string]]
    assignedTo*: Option[string]

  UpdateStatusRequest* = object
    status*: string

  TaskListResponse* = object
    tasks*: seq[Task]
    totalCount*: int

  ErrorResponse* = object
    error*: string

# Global task store with thread safety
var 
  taskStore = initTable[string, Task]()
  taskLock: Lock
  taskCounter = 0

initLock(taskLock)

proc generateTaskId(): string =
  withLock taskLock:
    inc taskCounter
    result = "task-" & $taskCounter

proc parseTaskStatus(s: string): Option[TaskStatus] =
  case s.toLowerAscii()
  of "pending": some(tsPending)
  of "in-progress", "in_progress": some(tsInProgress)
  of "completed": some(tsCompleted)
  of "cancelled": some(tsCancelled)
  else: none(TaskStatus)

proc parseTaskPriority(s: string): Option[TaskPriority] =
  case s.toLowerAscii()
  of "low": some(tpLow)
  of "medium": some(tpMedium)
  of "high": some(tpHigh)
  of "urgent": some(tpUrgent)
  else: none(TaskPriority)

proc taskToJson(task: Task): JsonNode =
  result = %* {
    "id": task.id,
    "title": task.title,
    "status": $task.status,
    "priority": $task.priority,
    "tags": task.tags,
    "created_at": task.createdAt,
    "updated_at": task.updatedAt
  }
  
  if task.description.isSome:
    result["description"] = %task.description.get()
  
  if task.assignedTo.isSome:
    result["assigned_to"] = %task.assignedTo.get()

proc createTask(req: CreateTaskRequest): Task =
  let now = epochTime()
  result = Task(
    id: generateTaskId(),
    title: req.title,
    description: req.description,
    status: tsPending,
    priority: if req.priority.isSome: 
                parseTaskPriority(req.priority.get()).get(tpMedium)
              else: 
                tpMedium,
    tags: if req.tags.isSome: req.tags.get() else: @[],
    assignedTo: req.assignedTo,
    createdAt: now,
    updatedAt: now
  )
  
  withLock taskLock:
    taskStore[result.id] = result

proc getTask(id: string): Option[Task] =
  withLock taskLock:
    if taskStore.hasKey(id):
      result = some(taskStore[id])
    else:
      result = none(Task)

proc updateTask(id: string, updates: UpdateTaskRequest): Option[Task] =
  withLock taskLock:
    if not taskStore.hasKey(id):
      return none(Task)
    
    var task = taskStore[id]
    
    if updates.title.isSome:
      task.title = updates.title.get()
    
    if updates.description.isSome:
      task.description = updates.description
    
    if updates.status.isSome:
      let status = parseTaskStatus(updates.status.get())
      if status.isSome:
        task.status = status.get()
    
    if updates.priority.isSome:
      let priority = parseTaskPriority(updates.priority.get())
      if priority.isSome:
        task.priority = priority.get()
    
    if updates.tags.isSome:
      task.tags = updates.tags.get()
    
    if updates.assignedTo.isSome:
      task.assignedTo = updates.assignedTo
    
    task.updatedAt = epochTime()
    taskStore[id] = task
    result = some(task)

proc updateTaskStatus(id: string, status: TaskStatus): Option[Task] =
  withLock taskLock:
    if not taskStore.hasKey(id):
      return none(Task)
    
    var task = taskStore[id]
    task.status = status
    task.updatedAt = epochTime()
    taskStore[id] = task
    result = some(task)

proc deleteTask(id: string): bool =
  withLock taskLock:
    if taskStore.hasKey(id):
      taskStore.del(id)
      result = true
    else:
      result = false

proc listTasks(statusFilter: Option[string] = none(string), 
               assignedToFilter: Option[string] = none(string)): seq[Task] =
  result = @[]
  
  withLock taskLock:
    for task in taskStore.values:
      var include = true
      
      if statusFilter.isSome:
        let status = parseTaskStatus(statusFilter.get())
        if status.isSome and task.status != status.get():
          include = false
      
      if assignedToFilter.isSome and include:
        if task.assignedTo.isNone or task.assignedTo.get() != assignedToFilter.get():
          include = false
      
      if include:
        result.add(task)
  
  # Sort by creation time (newest first)
  result.sort do (a, b: Task) -> int:
    if a.createdAt > b.createdAt: -1
    elif a.createdAt < b.createdAt: 1
    else: 0

# Initialize with sample data
proc initSampleData() =
  discard createTask(CreateTaskRequest(
    title: "Learn Nim",
    description: some("Master systems programming with Nim"),
    priority: some("high"),
    tags: some(@["nim", "learning", "systems"]),
    assignedTo: some("developer")
  ))
  
  discard createTask(CreateTaskRequest(
    title: "Build REST API",
    description: some("Create REST API with Jester framework"),
    priority: some("urgent"),
    tags: some(@["nim", "api", "rest", "jester"]),
    assignedTo: some("backend-team")
  ))
  
  discard createTask(CreateTaskRequest(
    title: "Write Tests",
    description: some("Add unit tests with unittest2"),
    priority: some("medium"),
    tags: some(@["testing", "quality"]),
    assignedTo: none(string)
  ))

# Initialize sample data on startup
initSampleData()

# Routes
routes:
  # Health check
  get "/health":
    resp %* {
      "status": "healthy",
      "service": "task-api",
      "version": "1.0.0",
      "timestamp": epochTime()
    }
  
  # List tasks
  get "/api/tasks":
    let statusFilter = 
      if request.params.hasKey("status"):
        some(request.params["status"])
      else:
        none(string)
    
    let assignedToFilter = 
      if request.params.hasKey("assigned_to"):
        some(request.params["assigned_to"])
      else:
        none(string)
    
    let tasks = listTasks(statusFilter, assignedToFilter)
    let response = TaskListResponse(
      tasks: tasks,
      totalCount: tasks.len
    )
    
    var jsonTasks = newJArray()
    for task in tasks:
      jsonTasks.add(taskToJson(task))
    
    resp %* {
      "tasks": jsonTasks,
      "total_count": tasks.len
    }
  
  # Get task by ID
  get "/api/tasks/@id":
    let taskOpt = getTask(@"id")
    if taskOpt.isSome:
      resp taskToJson(taskOpt.get())
    else:
      resp Http404, %* {"error": "Task not found"}
  
  # Create task
  post "/api/tasks":
    try:
      let body = parseJson(request.body)
      
      if not body.hasKey("title"):
        resp Http400, %* {"error": "Missing required field: title"}
        return
      
      let req = CreateTaskRequest(
        title: body["title"].getStr(),
        description: if body.hasKey("description"): 
                       some(body["description"].getStr()) 
                     else: 
                       none(string),
        priority: if body.hasKey("priority"): 
                    some(body["priority"].getStr()) 
                  else: 
                    none(string),
        tags: if body.hasKey("tags"):
                some(body["tags"].getElems().mapIt(it.getStr()))
              else:
                none(seq[string]),
        assignedTo: if body.hasKey("assigned_to"):
                      some(body["assigned_to"].getStr())
                    else:
                      none(string)
      )
      
      let task = createTask(req)
      resp Http201, taskToJson(task)
    except:
      resp Http400, %* {"error": "Invalid JSON"}
  
  # Update task
  put "/api/tasks/@id":
    try:
      let body = parseJson(request.body)
      let updates = UpdateTaskRequest(
        title: if body.hasKey("title"): 
                 some(body["title"].getStr()) 
               else: 
                 none(string),
        description: if body.hasKey("description"): 
                       some(body["description"].getStr()) 
                     else: 
                       none(string),
        status: if body.hasKey("status"): 
                  some(body["status"].getStr()) 
                else: 
                  none(string),
        priority: if body.hasKey("priority"): 
                    some(body["priority"].getStr()) 
                  else: 
                    none(string),
        tags: if body.hasKey("tags"):
                some(body["tags"].getElems().mapIt(it.getStr()))
              else:
                none(seq[string]),
        assignedTo: if body.hasKey("assigned_to"):
                      some(body["assigned_to"].getStr())
                    else:
                      none(string)
      )
      
      let taskOpt = updateTask(@"id", updates)
      if taskOpt.isSome:
        resp taskToJson(taskOpt.get())
      else:
        resp Http404, %* {"error": "Task not found"}
    except:
      resp Http400, %* {"error": "Invalid JSON"}
  
  # Update task status
  patch "/api/tasks/@id/status":
    try:
      let body = parseJson(request.body)
      
      if not body.hasKey("status"):
        resp Http400, %* {"error": "Missing required field: status"}
        return
      
      let statusStr = body["status"].getStr()
      let statusOpt = parseTaskStatus(statusStr)
      
      if statusOpt.isNone:
        resp Http400, %* {"error": "Invalid status value"}
        return
      
      let taskOpt = updateTaskStatus(@"id", statusOpt.get())
      if taskOpt.isSome:
        resp taskToJson(taskOpt.get())
      else:
        resp Http404, %* {"error": "Task not found"}
    except:
      resp Http400, %* {"error": "Invalid JSON"}
  
  # Delete task
  delete "/api/tasks/@id":
    if deleteTask(@"id"):
      resp Http204, ""
    else:
      resp Http404, %* {"error": "Task not found"}

when isMainModule:
  echo "🚀 Nim Task REST API Server"
  echo "📍 Listening on http://localhost:8080"
  echo "🔍 Health check: http://localhost:8080/health\n"

File: code/julia/rest/server/src/TaskServer.jl
module TaskServer

using HTTP
using JSON3
using StructTypes
using Dates
using UUIDs
using Sockets

# Type definitions
@enum TaskStatus begin
    PENDING
    IN_PROGRESS
    COMPLETED
    CANCELLED
end

@enum TaskPriority begin
    LOW
    MEDIUM
    HIGH
    URGENT
end

mutable struct Task
    id::String
    title::String
    description::Union{String, Nothing}
    status::TaskStatus
    priority::TaskPriority
    tags::Vector{String}
    assigned_to::Union{String, Nothing}
    created_at::Float64
    updated_at::Float64
end

struct CreateTaskRequest
    title::String
    description::Union{String, Nothing}
    priority::Union{String, Nothing}
    tags::Union{Vector{String}, Nothing}
    assigned_to::Union{String, Nothing}
end

struct UpdateTaskRequest
    title::Union{String, Nothing}
    description::Union{String, Nothing}
    status::Union{String, Nothing}
    priority::Union{String, Nothing}
    tags::Union{Vector{String}, Nothing}
    assigned_to::Union{String, Nothing}
end

struct UpdateStatusRequest
    status::String
end

struct TaskListResponse
    tasks::Vector{Task}
    total_count::Int
end

struct ErrorResponse
    error::String
end

# JSON serialization
StructTypes.StructType(::Type{Task}) = StructTypes.Mutable()
StructTypes.StructType(::Type{CreateTaskRequest}) = StructTypes.Struct()
StructTypes.StructType(::Type{UpdateTaskRequest}) = StructTypes.Struct()
StructTypes.StructType(::Type{UpdateStatusRequest}) = StructTypes.Struct()
StructTypes.StructType(::Type{TaskListResponse}) = StructTypes.Struct()
StructTypes.StructType(::Type{ErrorResponse}) = StructTypes.Struct()

# Custom JSON conversion for enums
function JSON3.write(io::IO, status::TaskStatus)
    status_str = if status == PENDING
        "pending"
    elseif status == IN_PROGRESS
        "in-progress"
    elseif status == COMPLETED
        "completed"
    else
        "cancelled"
    end
    JSON3.write(io, status_str)
end

function JSON3.write(io::IO, priority::TaskPriority)
    priority_str = if priority == LOW
        "low"
    elseif priority == MEDIUM
        "medium"
    elseif priority == HIGH
        "high"
    else
        "urgent"
    end
    JSON3.write(io, priority_str)
end

# Parse status from string
function parse_status(s::Union{String, Nothing})
    isnothing(s) && return nothing
    s_lower = lowercase(s)
    if s_lower == "pending"
        return PENDING
    elseif s_lower in ["in-progress", "in_progress"]
        return IN_PROGRESS
    elseif s_lower == "completed"
        return COMPLETED
    elseif s_lower == "cancelled"
        return CANCELLED
    else
        return nothing
    end
end

# Parse priority from string
function parse_priority(s::Union{String, Nothing})
    isnothing(s) && return nothing
    s_lower = lowercase(s)
    if s_lower == "low"
        return LOW
    elseif s_lower == "medium"
        return MEDIUM
    elseif s_lower == "high"
        return HIGH
    elseif s_lower == "urgent"
        return URGENT
    else
        return nothing
    end
end

# Global task store (thread-safe with lock)
const task_store = Dict{String, Task}()
const task_lock = ReentrantLock()
task_counter = Ref(0)

function generate_task_id()
    lock(task_lock) do
        task_counter[] += 1
        return "task-$(task_counter[])"
    end
end

function create_task(req::CreateTaskRequest)
    task_id = generate_task_id()
    now_time = time()
    
    priority = parse_priority(req.priority)
    if isnothing(priority)
        priority = MEDIUM
    end
    
    task = Task(
        task_id,
        req.title,
        req.description,
        PENDING,
        priority,
        isnothing(req.tags) ? String[] : req.tags,
        req.assigned_to,
        now_time,
        now_time
    )
    
    lock(task_lock) do
        task_store[task_id] = task
    end
    
    return task
end

function get_task(id::String)
    lock(task_lock) do
        return get(task_store, id, nothing)
    end
end

function update_task(id::String, updates::UpdateTaskRequest)
    lock(task_lock) do
        if !haskey(task_store, id)
            return nothing
        end
        
        task = task_store[id]
        
        if !isnothing(updates.title)
            task.title = updates.title
        end
        
        if !isnothing(updates.description)
            task.description = updates.description
        end
        
        if !isnothing(updates.status)
            status = parse_status(updates.status)
            if !isnothing(status)
                task.status = status
            end
        end
        
        if !isnothing(updates.priority)
            priority = parse_priority(updates.priority)
            if !isnothing(priority)
                task.priority = priority
            end
        end
        
        if !isnothing(updates.tags)
            task.tags = updates.tags
        end
        
        if !isnothing(updates.assigned_to)
            task.assigned_to = updates.assigned_to
        end
        
        task.updated_at = time()
        return task
    end
end

function update_task_status(id::String, status::TaskStatus)
    lock(task_lock) do
        if !haskey(task_store, id)
            return nothing
        end
        
        task = task_store[id]
        task.status = status
        task.updated_at = time()
        return task
    end
end

function delete_task(id::String)
    lock(task_lock) do
        if haskey(task_store, id)
            delete!(task_store, id)
            return true
        end
        return false
    end
end

function list_tasks(; status_filter=nothing, assigned_to_filter=nothing)
    lock(task_lock) do
        tasks = Task[]
        
        for task in values(task_store)
            include_task = true
            
            if !isnothing(status_filter)
                status = parse_status(status_filter)
                if !isnothing(status) && task.status != status
                    include_task = false
                end
            end
            
            if include_task && !isnothing(assigned_to_filter)
                if isnothing(task.assigned_to) || task.assigned_to != assigned_to_filter
                    include_task = false
                end
            end
            
            if include_task
                push!(tasks, task)
            end
        end
        
        # Sort by creation time (newest first)
        sort!(tasks, by=t -> -t.created_at)
        return tasks
    end
end

# Initialize with sample data
function init_sample_data()
    create_task(CreateTaskRequest(
        "Learn Julia",
        "Master scientific computing with Julia",
        "high",
        ["julia", "learning", "scientific"],
        "developer"
    ))
    
    create_task(CreateTaskRequest(
        "Build REST API",
        "Create REST API with HTTP.jl",
        "urgent",
        ["julia", "api", "rest"],
        "backend-team"
    ))
    
    create_task(CreateTaskRequest(
        "Write Tests",
        "Add unit tests with Test package",
        "medium",
        ["testing", "quality"],
        nothing
    ))
end

# HTTP Handlers
function health_handler(req::HTTP.Request)
    response = Dict(
        "status" => "healthy",
        "service" => "task-api",
        "version" => "1.0.0",
        "timestamp" => time()
    )
    return HTTP.Response(200, ["Content-Type" => "application/json"], JSON3.write(response))
end

function list_tasks_handler(req::HTTP.Request)
    query = HTTP.queryparams(HTTP.URI(req.target))
    status_filter = get(query, "status", nothing)
    assigned_to_filter = get(query, "assigned_to", nothing)
    
    tasks = list_tasks(status_filter=status_filter, assigned_to_filter=assigned_to_filter)
    response = TaskListResponse(tasks, length(tasks))
    
    return HTTP.Response(200, ["Content-Type" => "application/json"], JSON3.write(response))
end

function get_task_handler(req::HTTP.Request, id::String)
    task = get_task(id)
    
    if isnothing(task)
        error_response = ErrorResponse("Task not found")
        return HTTP.Response(404, ["Content-Type" => "application/json"], JSON3.write(error_response))
    end
    
    return HTTP.Response(200, ["Content-Type" => "application/json"], JSON3.write(task))
end

function create_task_handler(req::HTTP.Request)
    try
        body = String(req.body)
        json_data = JSON3.read(body)
        
        if !haskey(json_data, :title)
            error_response = ErrorResponse("Missing required field: title")
            return HTTP.Response(400, ["Content-Type" => "application/json"], JSON3.write(error_response))
        end
        
        create_req = CreateTaskRequest(
            json_data.title,
            get(json_data, :description, nothing),
            get(json_data, :priority, nothing),
            get(json_data, :tags, nothing),
            get(json_data, :assigned_to, nothing)
        )
        
        task = create_task(create_req)
        return HTTP.Response(201, ["Content-Type" => "application/json"], JSON3.write(task))
    catch e
        error_response = ErrorResponse("Invalid request: $(e)")
        return HTTP.Response(400, ["Content-Type" => "application/json"], JSON3.write(error_response))
    end
end

function update_task_handler(req::HTTP.Request, id::String)
    try
        body = String(req.body)
        json_data = JSON3.read(body)
        
        update_req = UpdateTaskRequest(
            get(json_data, :title, nothing),
            get(json_data, :description, nothing),
            get(json_data, :status, nothing),
            get(json_data, :priority, nothing),
            get(json_data, :tags, nothing),
            get(json_data, :assigned_to, nothing)
        )
        
        task = update_task(id, update_req)
        
        if isnothing(task)
            error_response = ErrorResponse("Task not found")
            return HTTP.Response(404, ["Content-Type" => "application/json"], JSON3.write(error_response))
        end
        
        return HTTP.Response(200, ["Content-Type" => "application/json"], JSON3.write(task))
    catch e
        error_response = ErrorResponse("Invalid request: $(e)")
        return HTTP.Response(400, ["Content-Type" => "application/json"], JSON3.write(error_response))
    end
end

function update_task_status_handler(req::HTTP.Request, id::String)
    try
        body = String(req.body)
        json_data = JSON3.read(body)
        
        if !haskey(json_data, :status)
            error_response = ErrorResponse("Missing required field: status")
            return HTTP.Response(400, ["Content-Type" => "application/json"], JSON3.write(error_response))
        end
        
        status = parse_status(json_data.status)
        if isnothing(status)
            error_response = ErrorResponse("Invalid status value")
            return HTTP.Response(400, ["Content-Type" => "application/json"], JSON3.write(error_response))
        end
        
        task = update_task_status(id, status)
        
        if isnothing(task)
            error_response = ErrorResponse("Task not found")
            return HTTP.Response(404, ["Content-Type" => "application/json"], JSON3.write(error_response))
        end
        
        return HTTP.Response(200, ["Content-Type" => "application/json"], JSON3.write(task))
    catch e
        error_response = ErrorResponse("Invalid request: $(e)")
        return HTTP.Response(400, ["Content-Type" => "application/json"], JSON3.write(error_response))
    end
end

function delete_task_handler(req::HTTP.Request, id::String)
    if delete_task(id)
        return HTTP.Response(204, ["Content-Type" => "application/json"], "")
    else
        error_response = ErrorResponse("Task not found")
        return HTTP.Response(404, ["Content-Type" => "application/json"], JSON3.write(error_response))
    end
end

# Router
function router(req::HTTP.Request)
    uri = HTTP.URI(req.target)
    path = uri.path
    method = req.method
    
    # Health check
    if path == "/health" && method == "GET"
        return health_handler(req)
    end
    
    # Task routes
    if startswith(path, "/api/tasks")
        path_parts = split(path, "/"; keepempty=false)
        
        if length(path_parts) == 2  # /api/tasks
            if method == "GET"
                return list_tasks_handler(req)
            elseif method == "POST"
                return create_task_handler(req)
            end
        elseif length(path_parts) == 3  # /api/tasks/{id}
            task_id = path_parts[3]
            if method == "GET"
                return get_task_handler(req, task_id)
            elseif method == "PUT"
                return update_task_handler(req, task_id)
            elseif method == "DELETE"
                return delete_task_handler(req, task_id)
            end
        elseif length(path_parts) == 4 && path_parts[4] == "status"  # /api/tasks/{id}/status
            task_id = path_parts[3]
            if method == "PATCH"
                return update_task_status_handler(req, task_id)
            end
        end
    end
    
    # 404 for unmatched routes
    error_response = ErrorResponse("Not found")
    return HTTP.Response(404, ["Content-Type" => "application/json"], JSON3.write(error_response))
end

function start_server(port=8080)
    # Initialize sample data
    init_sample_data()
    
    println("🚀 Julia Task REST API Server")
    println("📍 Listening on http://localhost:$port")
    println("🔍 Health check: http://localhost:$port/health\n")
    
    HTTP.serve(router, "0.0.0.0", port)
end

end # module